private-identity-audit/src/proofs.rs

```diff
@@ -17,26 +17,150 @@
+use sp_std::vec::Vec;
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct InitialMessage {
     a: RistrettoPoint,
     pub generators: Vec<RistrettoPoint>,
 }
 
-#[derive(Clone)]
+impl Encode for InitialMessage {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        32 + self.generators.len() * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let a = self.a.compress();
+        let generators: Vec<[u8; 32]> = self
+            .generators
+            .clone()
+            .into_iter()
+            .map(|g| *g.compress().as_bytes())
+            .collect::<Vec<_>>();
+
+        a.as_bytes().encode_to(dest);
+        generators.encode_to(dest);
+    }
+}
+
+impl Decode for InitialMessage {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let (a, generators) = <([u8; 32], Vec<[u8; 32]>)>::decode(input)?;
+        let a = CompressedRistretto(a)
+            .decompress()
+            .ok_or_else(|| CodecError::from("InitialMessage `a` point is invalid"))?;
+
+        let generators: Vec<RistrettoPoint> = generators
+            .into_iter()
+            .map(|g| {
+                CompressedRistretto(g)
+                    .decompress()
+                    .unwrap_or_else(RistrettoPoint::default)
+            })
+            .collect::<Vec<_>>();
+
+        Ok(InitialMessage { a, generators })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct FinalResponse {
-    s: Vec<Scalar>,
+    response: Vec<Scalar>,
 }
 
-#[derive(Clone)]
+impl Encode for FinalResponse {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        self.response.len() * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let response: Vec<[u8; 32]> = self
+            .response
+            .clone()
+            .into_iter()
+            .map(|g| *g.as_bytes())
+            .collect::<Vec<_>>();
+
+        response.encode_to(dest);
+    }
+}
+
+impl Decode for FinalResponse {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let response = <Vec<[u8; 32]>>::decode(input)?;
+
+        let s: Vec<Scalar> = response
+            .into_iter()
+            .map(Scalar::from_bits)
+            .collect::<Vec<_>>();
+
+        Ok(FinalResponse { response: s })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct Secrets {
     rands: Vec<Scalar>,
     secrets: Vec<Scalar>,
 }
 
+impl Encode for Secrets {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        self.rands.len() * 32 + self.secrets.len() * 32
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let rands: Vec<[u8; 32]> = self
+            .rands
+            .clone()
+            .into_iter()
+            .map(|g| *g.as_bytes())
+            .collect::<Vec<_>>();
```

Comment:
Performance: use references.
```suggestion
        let rands = self
            .rands
            .iter()
            .map(|g| g.as_bytes())
            .collect::<Vec<_>>();
```